portfolio/src/main/webapp/script.js

```diff
@@ -15,14 +15,26 @@
 /**
  * Adds a random greeting to the page.
  */
-function addRandomGreeting() {
-  const greetings =
-      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!'];
 
+var used_greetings = [];
+var greetings =
+      ['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!', 'Hallo Welt!', 'こんにちは世界', '안녕하세요 세계!', 'Привет мир!'];
```

Comment:
This is a clever twist, to not reuse greetings until they've all been seen.

Let me suggest a slight modification of the structure, since it doesn't seem like used_greetings gets used for anything other than resetting the greetings list when all greetings have been used. This requires mutating both the used_greetings (by adding the value) as well as the greetings variable in order to work properly. If, for some reason, there was a code path that didn't update the used_greetings properly, then some of the initial values might get lost. And this way the only state that needs to be updated is active_greetings, instead of both greetings and used_greetings.

So, what about this:
var all_greetings = 
['Hello world!', '¡Hola Mundo!', '你好，世界！', 'Bonjour le monde!', 'Hallo Welt!', 'こんにちは世界', '안녕하세요 세계!', 'Привет мир!'];
var active_greetings = all_greetings;

Then you can drop lines 26-28 which have
  used_greetings.push(greeting);

And then simplify the update and reset code to be this:
active_greetings.splice(active_greetings.indexOf(greeting), 1);
  if (active_greetings.length == 0) {
      active_greetings = all_greetings;
  }

Reply:
Great idea! I incorporated these changes. I found out that `var active_greetings = all_greetings` only copied the array reference, leading to incorrect behavior, so I made an additional change to make a true copy of `all_greetings` by doing `var active_greetings = all_greetings.splice()` instead.